File: ios/rCaltrain/Trip.swift
//
//  Trip.swift
//  rCaltrain
//
//  Created by Wanzhang Sheng on 10/23/14.
//  Copyright (c) 2014-2015 Ranmocy. All rights reserved.
//

import Foundation

class Trip {

    let id : String
    var stops = [Stop]()

    init(id: String) {
        self.id = id
    }

    convenience init(id: String, stopsArray: NSArray) {
        self.init(id: id)

        // FIXME: can't use `for data in stopsArray as [NSArray]`
        // which will crash when compile as release version
        for data in stopsArray {
            assert(data.count == 2, "data length is \(data.count), expected 2!")

            let stationId = data[0] as Int;
            let time = NSDate(timeIntervalSince1970: NSTimeInterval(data[1] as Int))
            
            if let station = Station.getStation(byId: stationId) {
                self.stops.append(Stop(station: station, departureTime: time, arrivalTime: time))
            } else {
                fatalError("can't find station id\(stationId)")
            }
        }
    }

    func addStop(stop : Stop) -> Trip {
        self.stops.append(stop)
        return self
    }

    func findFrom(from: Station, to: Station) -> (Stop, Stop)? {
        var i: Int = 0
        var fromStop: Stop?, toStop: Stop?

        // find the departure stop
        while (i < stops.count){
            if (stops[i].station === from) {
                fromStop = stops[i]
                break
            }
            i++
        }

        // if missing
        if (fromStop == nil) {
            return nil
        }

        // from and to can't be the same
        i++

        // find the arrival stop
        while (i < stops.count) {
            if (stops[i].station === to) {
                toStop = stops[i]
                break
            }
            i++
        }

        // if missing
        if (toStop == nil) {
            return nil
        }
        
        return (fromStop!, toStop!)
    }

}

File: android/app/src/main/java/me/ranmocy/rcaltrain/Trip.java
package me.ranmocy.rcaltrain;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

/**
 * Trip model.
 */
public class Trip {

    public class Stop {

        private final Station station;
        private final Date time;

        private Stop(Station station, Date time) {
            this.station = station;
            this.time = time;
        }

        public Date getTime() {
            return time;
        }
    }

    private String id;
    private Service service;
    private final List<Stop> stops = new ArrayList<>();

    public Trip(String id, Service service) {
        this.id = id;
        this.service = service;
    }

    public void addStop(Station station, Date time) {
        stops.add(new Stop(station, time));
    }

    public List<Stop> getStopList() {
        return stops;
    }

    public List<Station> getStationList() {
        List<Station> stations = new ArrayList<>();
        for (Stop stop : stops) {
            stations.add(stop.station);
        }
        return stations;
    }
}
